### ISSUE TITLE
RetryOption.setMaximumAttempts(0) seems broken for LocalActivities

### ISSUE THREAD

**Issue by ghaskins** [temporalio/sdk-java#2254]
## Expected Behavior

The default RetryOptions or explicitly setting RetryOption.setMaximumAttempts(0) for a local activity is suggested to mean "unlimited retries"

## Actual Behavior

Local activities will not retry with default RetryOptions or explicitly setting RetryOptions.setMaxiumAttempts(0).  Instead, they will error out after 0 retry attempts.  This is inconsistent with the documentation, and I believe it is also inconsistent with normal activitiy behavior

## Steps to Reproduce the Problem

 Run the following code snippet modified from the HelloActivity sample.  The primary changes are
 
 1) We now explicitly set RetryOption.setMaximumAttempts(0)
 2) We are calling a LocalActivity instead of a normal activity
 3) The activity intentionally sleeps as to trigger the StartToClose timeout
 
 If you run this code as is, it will fail after 1 second.  The expected behavior is that it would run indefinitely due to "unlimited" retries.
 
 You can confirm that the general test structure is suited to reproduction by changing the setMaximumAttempts to a positive integer (e.g. 10) and confirming that it does retry N times when it is positive.  However, the default/0 case for unlimited does not seem to work.
 
 ```
 /*
 *  Copyright (c) 2020 Temporal Technologies, Inc. All Rights Reserved
 *
 *  Copyright 2012-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Modifications copyright (C) 2017 Uber Technologies, Inc.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not
 *  use this file except in compliance with the License. A copy of the License is
 *  located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 *  or in the "license" file accompanying this file. This file is distributed on
 *  an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *  express or implied. See the License for the specific language governing
 *  permissions and limitations under the License.
 */

package org.example;

import io.temporal.activity.ActivityInterface;
import io.temporal.activity.ActivityMethod;
import io.temporal.activity.ActivityOptions;
import io.temporal.activity.LocalActivityOptions;
import io.temporal.client.WorkflowClient;
import io.temporal.client.WorkflowOptions;
import io.temporal.common.RetryOptions;
import io.temporal.serviceclient.WorkflowServiceStubs;
import io.temporal.worker.Worker;
import io.temporal.worker.WorkerFactory;
import io.temporal.workflow.Workflow;
import io.temporal.workflow.WorkflowInterface;
import io.temporal.workflow.WorkflowMethod;
import java.time.Duration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/** Sample Temporal Workflow Definition that executes a single Activity. */
public class HelloActivity {

    // Define the task queue name
    static final String TASK_QUEUE = "HelloActivityTaskQueue";

    // Define our workflow unique id
    static final String WORKFLOW_ID = "HelloActivityWorkflow";

    /**
     * The Workflow Definition's Interface must contain one method annotated with @WorkflowMethod.
     *
     * <p>Workflow Definitions should not contain any heavyweight computations, non-deterministic
     * code, network calls, database operations, etc. Those things should be handled by the
     * Activities.
     *
     * @see io.temporal.workflow.WorkflowInterface
     * @see io.temporal.workflow.WorkflowMethod
     */
    @WorkflowInterface
    public interface GreetingWorkflow {

        /**
         * This is the method that is executed when the Workflow Execution is started. The Workflow
         * Execution completes when this method finishes execution.
         */
        @WorkflowMethod
        String getGreeting(String name);
    }

    /**
     * This is the Activity Definition's Interface. Activities are building blocks of any Temporal
     * Workflow and contain any business logic that could perform long running computation, network
     * calls, etc.
     *
     * <p>Annotating Activity Definition methods with @ActivityMethod is optional.
     *
     * @see io.temporal.activity.ActivityInterface
     * @see io.temporal.activity.ActivityMethod
     */
    @ActivityInterface
    public interface GreetingActivities {

        // Define your activity method which can be called during workflow execution
        @ActivityMethod(name = "greet")
        String composeGreeting(String greeting, String name);
    }

    // Define the workflow implementation which implements our getGreeting workflow method.
    public static class GreetingWorkflowImpl implements GreetingWorkflow {

        /**
         * Define the GreetingActivities stub. Activity stubs are proxies for activity invocations that
         * are executed outside of the workflow thread on the activity worker, that can be on a
         * different host. Temporal is going to dispatch the activity results back to the workflow and
         * unblock the stub as soon as activity is completed on the activity worker.
         *
         * <p>In the {@link ActivityOptions} definition the "setStartToCloseTimeout" option sets the
         * overall timeout that our workflow is willing to wait for activity to complete. For this
         * example it is set to 2 seconds.
         */
        private final GreetingActivities activities =
                Workflow.newLocalActivityStub(
                        GreetingActivities.class,
                        LocalActivityOptions.newBuilder()
                                .setStartToCloseTimeout(Duration.ofSeconds(1))
                                .setRetryOptions(RetryOptions.newBuilder().setMaximumAttempts(0).build())
                                .build());

        @Override
        public String getGreeting(String name) {
            return activities.composeGreeting("Hello", name);
        }
    }

    /** Simple activity implementation, that concatenates two strings. */
    static class GreetingActivitiesImpl implements GreetingActivities {
        private static final Logger log = LoggerFactory.getLogger(GreetingActivitiesImpl.class);

        @Override
        public String composeGreeting(String greeting, String name) {
            log.info("Composing greeting...");

            System.out.println("getGreeting");
            try {
                Thread.sleep(1000000000);
            } catch (InterruptedException e) {
                throw new RuntimeException(e);
            }

            return greeting + " " + name + "!";
        }
    }

    /**
     * With our Workflow and Activities defined, we can now start execution. The main method starts
     * the worker and then the workflow.
     */
    public static void main(String[] args) {

        // Get a Workflow service stub.
        WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();

        /*
         * Get a Workflow service client which can be used to start, Signal, and Query Workflow Executions.
         */
        WorkflowClient client = WorkflowClient.newInstance(service);

        /*
         * Define the workflow factory. It is used to create workflow workers for a specific task queue.
         */
        WorkerFactory factory = WorkerFactory.newInstance(client);

        /*
         * Define the workflow worker. Workflow workers listen to a defined task queue and process
         * workflows and activities.
         */
        Worker worker = factory.newWorker(TASK_QUEUE);

        /*
         * Register our workflow implementation with the worker.
         * Workflow implementations must be known to the worker at runtime in
         * order to dispatch workflow tasks.
         */
        worker.registerWorkflowImplementationTypes(GreetingWorkflowImpl.class);

        /*
         * Register our Activity Types with the Worker. Since Activities are stateless and thread-safe,
         * the Activity Type is a shared instance.
         */
        worker.registerActivitiesImplementations(new GreetingActivitiesImpl());

        /*
         * Start all the workers registered for a specific task queue.
         * The started workers then start polling for workflows and activities.
         */
        factory.start();

        // Create the workflow client stub. It is used to start our workflow execution.
        GreetingWorkflow workflow =
                client.newWorkflowStub(
                        GreetingWorkflow.class,
                        WorkflowOptions.newBuilder()
                                .setWorkflowId(WORKFLOW_ID)
                                .setTaskQueue(TASK_QUEUE)
                                .setRetryOptions(RetryOptions.newBuilder().setMaximumAttempts(1).build())
                                .build());

        /*
         * Execute our workflow and wait for it to complete. The call to our getGreeting method is
         * synchronous.
         *
         * See {@link io.temporal.samples.hello.HelloSignal} for an example of starting workflow
         * without waiting synchronously for its result.
         */
        String greeting = workflow.getGreeting("World");

        // Display workflow execution results
        System.out.println(greeting);
        System.exit(0);
    }
}
```

**Comment by Quinn-With-Two-Ns** [comment]
Duplicate of https://github.com/temporalio/sdk-java/issues/1727
